### DISCUSSION TITLE
Problem using base64 encoded images as background or img src

### DISCUSSION THREAD

**Discussion by Bohemme** [mpdf/mpdf#1743] [Q&A] [answered]
Hi everyone:

Some years ago another user (futape) posted the same issue: #1104

Since mPDF 7.1.8 or since commit https://github.com/mpdf/mpdf/commit/20ff6399433c18233f31817ba2f35a86dd9d5e22 (PR https://github.com/mpdf/mpdf/pull/950) it's no longer possible to use a data:// URI as source of an image to draw. This is that commit fixe issue https://github.com/mpdf/mpdf/issues/949. Unfortunately instrad of implementing a blacklist disabling the phar:// wrapper, a whitelist only allowing http, https and file URIs as well as traditional paths has been implemented disabling any data:// URIs (as well as any other wrapper).

```php
<?php
$mpdf = new \Mpdf\Mpdf();
$mpdf->Image('data://image/jpg;base64,<base64-encoded jpeg>');
```
The documentation says otherwise, but I understand that it's an important security issue so it must be solved, like user jakejackson1 said:

You can work around this by configuring a whitelist you are comfortable with on your host when initialising mPDF

And it seems to solve the issue, but neither of them said exactly what needs to be done.
I've seen that in the ConfigVariables.php file there is an array of valid php streams:
```php
'whitelistStreamWrappers' => ['http', 'https', 'file', 'jpeg'],
```
but I don't know what should I add and where. I've tried adding `"data"`, the whole route of my php file, the name of the ajax stream, and so on with no good results.

Any help, please?

**Comment by finwe** [comment, marked as answer]
Show *your* failing code example.

**Reply by Bohemme** [reply]
Ok, in my php script I recieve a base64 string, which is correct as I can use any online base64 decoder and shows the image perfectly, or I can even write an html file with just 
`<img src="<?= $base64img?>">`
 and appears correctly in the explorer.
But mPdf blocks apparently that stream for security reasons and the exported pdf shows no image at all. Something like:
`#myheader{
		background-repeat: no-repeat;
		background-size: contain;
		background-position:center center;
		background-color: gray;
		background-image:  url(' . $budgetHeader . ');
	}
`
being $budgetHeader a correct base64 string, it just doesn't work.
If (as I'm doing now) I create an image file "on the fly" with that same string, the image file is generated perfectly using a function like this:

```php
function returnFile($data, $alt = '') {
	global $localDir;
	$name = time();
	$data = explode(',', $data);
	$firstPart = explode(';', $data[0]);
	$secondPart = explode('/', $firstPart[0]);
	$ext = str_replace('jpeg', 'jpg', $secondPart[1]);
	$imageFileName = $localDir . 'php/temp/' . $name . $alt . '.' . $ext;
	$file = fopen($imageFileName, "wb");
	fwrite($file, base64_decode($data[1]));
	fclose($file);
	return $imageFileName;
}
echo '<img src="'.returnFile($budgetHeader).'">
```
but I'd like not having to create that physical file (though I delete it once the pdf is generated)

**Reply by finwe** [reply]
what is the content of `$base64img` and/or `$budgetHeader` variable? or how is it created?

**Reply by Bohemme** [reply]
I have this webapp. There, I have an input field where the user can upload an image. I use a cropper jquery library so the final field has a element with the encoded base64 cropped pic. This works just fine.
Then the user can process this form and this image (just the base64 string) is passed via AJAX to a PHP script. There, I create the pdf where there some images (standard jpg files) are perfectly included but this base64 string, when inserted as a background-image property, doesn't appear.
As I've said, this base64 string is perfectly formed and can be used in a plain html file and renders the way is supposed to.
But mPDF, for security reasons (that "phar" thing I don't understand) blocks that string and can't be used as background-image or img src. 
The solution appears to have something to do with that whitelistStreamWrappers parameter (as that user said in the other thread) but I don't know how to make it work or what option to use there to avoid that security block.

**Reply by Bohemme** [reply]
$base64img or $budgetHeader are (or could be) the same string which is something like: 
'data:image/jpeg;charset=utf-8;base64,/9j/4AAQSkZJRgABAQEAYABgAAD//gA+Q1JFQVRPUjogZ2QtanBlZyB 2MS4wICh1c2luZyBJSkcgSlBFRyB2OTApLCBkZWZhd...........'

**Reply by finwe** [reply]
Try removing the charset part, it seems that mPDF doesn't expect it, given that the content is base64-encoded binary string

**Reply by finwe** [reply]
If that helps, I'll improve the data URI parsing to account for the charset definition

**Reply by Bohemme** [reply]
Well, that did it!! Withour further modifications, all images show up correctly, as background-images style or as img html element. Thanks a lot, you've made my day! Regards and great job!

**Reply by finwe** [reply]
and, whitelistStreamWrappers have no effect in this case given that the data URI is not considered a protocol and/or a wrapper.
